File: sdk/AuthSys-Device-Example/lua/AuthSys-Device.lua
local http = require("socket.http")
local ltn12 = require("ltn12")
local json = require("json")

local Device = {}
Device.__index = Device

function Device:new(appSecret, baseUrl)
    local obj = {
        appSecret = appSecret,
        baseUrl = (baseUrl or "https://authsys-main-production.up.railway.app/device"):gsub("/+$", ""),
        lastError = "",
        lastResponse = ""
    }
    setmetatable(obj, self)
    return obj
end

local function getHWID()
    local tmp = os.tmpname()
    os.execute('wmic bios get serialnumber > "' .. tmp .. '" 2>nul')
    local f = io.open(tmp, "r")
    if f then
        local content = f:read("*a")
        f:close()
        os.remove(tmp)
        local sn = content:gsub("SerialNumber", ""):gsub("%s+", "")
        if sn and sn ~= "" then return sn end
    end
    local f = io.open("/etc/machine-id", "r")
    if f then
        local id = f:read("*l")
        f:close()
        if id and id ~= "" then return id end
    end
    return "unknown"
end

local function postRequest(self, endpoint, jsonData)
    local url = self.baseUrl .. "/" .. endpoint
    local responseBody = {}
    local _, status = http.request{
        url = url,
        method = "POST",
        headers = { ["Content-Type"] = "application/json" },
        source = ltn12.source.string(jsonData),
        sink = ltn12.sink.table(responseBody)
    }
    if status ~= 200 then
        self.lastError = "HTTP " .. tostring(status)
        return "{}"
    end
    return table.concat(responseBody)
end

function Device:check()
    self.lastError = ""
    local hwid = getHWID()
    local body = '{"group_secret":"' .. self.appSecret .. '","hwid":"' .. hwid .. '"}'
    self.lastResponse = postRequest(self, "check", body)
    local ok, data = pcall(json.decode, self.lastResponse)
    if ok and data then
        if data.active == true then return true end
        self.lastError = data.message or "Device deactivated by admin"
        return false
    end
    self.lastError = "Failed to parse response"
    return false
end

function Device:register(deviceName)
    self.lastError = ""
    local hwid = getHWID()
    local body = '{"group_secret":"' .. self.appSecret .. '","hwid":"' .. hwid .. '"'
    if deviceName and deviceName ~= "" then
        body = body .. ',"device_name":"' .. deviceName .. '"'
    end
    body = body .. "}"
    self.lastResponse = postRequest(self, "register", body)
    local ok, data = pcall(json.decode, self.lastResponse)
    if ok and data then
        return data.active == true
    end
    return false
end

return Device

File: sdk/AuthSys-Device-Example/python/AuthSys-Device.py
import requests
import subprocess
import platform


class Device:
    def __init__(self, group_secret: str, base_url: str = "https://api.authsys.dpdns.org/api/v1/client"):
        self.group_secret = group_secret
        self.base_url = base_url.rstrip("/")
        self.last_error = ""
        self.last_response = ""

    @staticmethod
    def _get_hwid() -> str:
        try:
            if platform.system() == "Windows":
                output = subprocess.check_output(
                    ["wmic", "bios", "get", "serialnumber"], timeout=5
                ).decode().strip().split("\n")[1].strip()
                return output if output else "unknown"
            else:
                with open("/etc/machine-id") as f:
                    return f.read().strip()
        except Exception:
            return "unknown"

    def check(self) -> bool:
        self.last_error = ""
        try:
            hwid = self._get_hwid()
            payload = {"group_secret": self.group_secret, "hwid": hwid}
            r = requests.post(
                f"{self.base_url}/check",
                json=payload,
                timeout=15,
            )
            self.last_response = r.text
            data = r.json()
            if data.get("active") is True:
                return True
            self.last_error = data.get("message", "Device deactivated by admin")
            return False
        except Exception as e:
            self.last_error = str(e)
            return False

    def register(self, device_name: str = "") -> bool:
        self.last_error = ""
        try:
            hwid = self._get_hwid()
            payload = {"group_secret": self.group_secret, "hwid": hwid}
            if device_name:
                payload["device_name"] = device_name
            r = requests.post(
                f"{self.base_url}/register",
                json=payload,
                timeout=15,
            )
            self.last_response = r.text
            data = r.json()
            return data.get("active") is True
        except Exception as e:
            self.last_error = str(e)
            return False

